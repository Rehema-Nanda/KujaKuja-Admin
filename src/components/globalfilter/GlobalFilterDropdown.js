import React from "react";

import {
    Popover, PopoverBody, Badge,
} from "reactstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import PropTypes, { arrayOf } from "prop-types";

import GlobalFilterCheckbox from "./GlobalFilterCheckbox";
import GlobalFilterDropdownSearch from "./GlobalFilterDropdownSearch";

export default class GlobalFilterDropdown extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            popoverOpen: false,
            selectedItemName: "",
            itemsToDisplay: props.items,
            searchTerm: "",
        };

        this.toggle = this.toggle.bind(this);
    }

    componentDidUpdate = (prevProps) => {
        const { items, selectedItems } = this.props;

        if (prevProps.items !== items) {
            this.setState({
                itemsToDisplay: items,
                searchTerm: "",
            });
        }
        if (prevProps.selectedItems !== selectedItems) {
            this.getSelectedItemName(selectedItems);
        }
    }

    toggle = () => {
        const { popoverOpen } = this.state;

        this.setState(
            {
                popoverOpen: !popoverOpen,
            },
        );
    };

    getSelectedItemName = (selectedItemIds) => {
        const { items, itemLabelProperty } = this.props;

        if (selectedItemIds.length !== 1) {
            return;
        }

        const selectedItem = items.find((item) => {
            return item.id === selectedItemIds[0];
        });

        if (selectedItem) {
            this.setState({ selectedItemName: selectedItem[itemLabelProperty] });
        }
    }

    truncate = (str, id) => {
        let length = 8;

        if (id === "ServicePointsDropDown") {
            length += 19;
        }
        else if (id === "UsersDropDown") {
            length += 5;
        }

        if (str.length <= length) {
            return str;
        }
        return `${str.substring(0, length)}...`;
    }

    setItemsToDisplay = (items) => {
        this.setState({
            itemsToDisplay: items,
        });
    };

    setSearchTerm = (searchTerm) => {
        this.setState({
            searchTerm: searchTerm,
        });
    };

    render() {
        const { popoverOpen, selectedItemName, itemsToDisplay, searchTerm } = this.state;
        const {
            id,
            items,
            title,
            selectedItems,
            selectItemHandler,
            itemLabelProperty,
            selectItemHandlerArgs,
            showSearch,
            isServicePoint,
        } = this.props;

        return (
            <div>
                <div id={id} onClick={this.toggle} className="map-controls-button">
                    <Badge color="secondary" pill>
                        {
                            selectedItems.length > 0
                                ? selectedItems.length
                                : "All"
                        }
                    </Badge>
                    &nbsp;
                    {selectedItems.length === 1
                        ? this.truncate(selectedItemName, id)
                        : title}
                    <div className="map-controls-button-caret"><FontAwesomeIcon icon="caret-down" /></div>
                </div>
                <Popover placement="bottom" isOpen={popoverOpen} target={id} toggle={this.toggle} className="map-controls-popover">
                    {showSearch
                    && (
                        <GlobalFilterDropdownSearch
                            items={items}
                            itemLabelProperty={itemLabelProperty}
                            setResults={this.setItemsToDisplay}
                            searchTerm={searchTerm}
                            setSearchTerm={this.setSearchTerm}
                        />
                    )}
                    <PopoverBody>
                        <ul className="map-controls-popover-list">
                            {itemsToDisplay.map((item) => {
                                return (
                                    <li key={item.id}>
                                        <GlobalFilterCheckbox
                                            item={item}
                                            itemLabelProperty={itemLabelProperty}
                                            selectItemHandler={selectItemHandler}
                                            selectItemHandlerArgs={selectItemHandlerArgs}
                                            selectedItems={selectedItems}
                                            isServicePoint={isServicePoint}
                                        />
                                    </li>
                                );
                            })}
                        </ul>
                    </PopoverBody>
                </Popover>
            </div>
        );
    }
}

GlobalFilterDropdown.propTypes = {
    selectedItems: arrayOf(PropTypes.any).isRequired,
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    itemLabelProperty: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    selectItemHandler: PropTypes.func.isRequired,
    selectItemHandlerArgs: PropTypes.arrayOf(PropTypes.string).isRequired,
    showSearch: PropTypes.bool,
    isServicePoint: PropTypes.bool,
};

GlobalFilterDropdown.defaultProps = {
    isServicePoint: false,
    showSearch: false,
};
