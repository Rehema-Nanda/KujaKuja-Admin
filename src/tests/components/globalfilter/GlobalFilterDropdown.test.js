import React from "react";
import { shallow } from "enzyme";
import GlobalFilterDropdown from "../../../components/globalfilter/GlobalFilterDropdown";

describe("<GlobalFilterDropdown />", () => {
    const testProps = {
        selectedItems: ["1"],
        id: "ServicePointsDropDown",
        items: [{ id: "2", name: "test" }],
        itemLabelProperty: "name",
    };
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<GlobalFilterDropdown {...testProps} />);
    });

    it("renders correctly", () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it("sets selecteditemname state when selectedItem prop changes", () => {
        wrapper.setProps({ selectedItems: ["2"] });
        expect(wrapper.state().selectedItemName).toEqual("test");
    });

    it("toggles the state of popoverOpen when the div is clicked", () => {
        const div = wrapper.find("div.map-controls-button");
        const state = wrapper.state().popoverOpen;
        div.simulate("click");
        expect(wrapper.state().popoverOpen).toEqual(!state);
    });
});
