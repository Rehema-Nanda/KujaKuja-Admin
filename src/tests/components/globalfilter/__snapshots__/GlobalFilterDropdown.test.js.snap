// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalFilterDropdown /> renders correctly 1`] = `
"<div>
  <div id=\\"ServicePointsDropDown\\" onClick={[Function: bound ]} className=\\"map-controls-button\\">
    <Badge color=\\"secondary\\" pill={true} tag=\\"span\\">
      1
    </Badge>
    Â 
    <div className=\\"map-controls-button-caret\\">
      <FontAwesomeIcon icon=\\"caret-down\\" />
    </div>
  </div>
  <mockConstructor placement=\\"bottom\\" isOpen={false} target=\\"ServicePointsDropDown\\" toggle={[Function: bound ]} className=\\"map-controls-popover\\">
    <PopoverBody tag=\\"div\\">
      <ul className=\\"map-controls-popover-list\\">
        <li>
          <GlobalFilterCheckbox item={{...}} itemLabelProperty=\\"name\\" selectItemHandler={[undefined]} selectItemHandlerArgs={[undefined]} selectedItems={{...}} isServicePoint={false} />
        </li>
      </ul>
    </PopoverBody>
  </mockConstructor>
</div>"
`;
